### PR TITLE
Adding more augmentation

### PR DESCRIPTION
**Pull request by srbhchandra**
Changes in this PR: 
 - adding more image augmentation techniques in `text_recognizer/stems/paragraph.py`

Wandb run 1 with `lr=0.0001`: https://wandb.ai/fullstackdeeplearning/fsdl-text-recognizer-2022-training/runs/2trhtbd4
Report for this run: https://wandb.ai/fullstackdeeplearning/fsdl-text-recognizer-2022-training/reports/Report-for-lilac-gorge-6--VmlldzoyMzgwNzQ1/edit

Wandb run 2 with `lr=0.001`: https://wandb.ai/fullstackdeeplearning/fsdl-text-recognizer-2022-training/runs/26zd6ttn
Training failed due to `nan` loss.

PS: I tried auto-tuning `lr` and it set `lr=0.00012022644346174131`. So didn't proceed since it is very close to what we anyway set (`lr=0.0001`)

Signed-off-by: srbhchandra <srbhchandra@gmail.com>

### CODE REVIEW

File: text_recognizer/data/util.py [hunk 1]
```diff
@@ -94,24 +94,24 @@ def split_dataset(base_dataset: BaseDataset, fraction: float, seed: int) -> Tupl
...
+#             transforms.RandomAffine(
+#                 degrees=1,
+#                 shear=(-10, 10),
+#                 resample=Image.BILINEAR,
+#             ),
+#         ]
+#     else:
+#         transforms_list = [transforms.CenterCrop(image_shape)]  # pad image to image_shape with 0
+#     transforms_list.append(transforms.ToTensor())
+#     return transforms.Compose(transforms_list)
```
Comment (srbhchandra) [inline text_recognizer/data/util.py:114]: just so we are sure of not using this code

File: text_recognizer/data/util.py [hunk 2]
```diff
@@ -94,26 +93,6 @@ def split_dataset(base_dataset: BaseDataset, fraction: float, seed: int) -> Tupl
...
-            transforms.RandomAffine(
-                degrees=1,
-                shear=(-10, 10),
-                resample=Image.BILINEAR,
-            ),
-        ]
-    else:
-        transforms_list = [transforms.CenterCrop(image_shape)]  # pad image to image_shape with 0
-    transforms_list.append(transforms.ToTensor())
-    return transforms.Compose(transforms_list)
```
Comment (srbhchandra) [inline text_recognizer/data/util.py:114]: removing to avoid confusion since it was not called anywhere

File: text_recognizer/stems/paragraph.py [hunk 3]
```diff
@@ -18,28 +18,51 @@ class ParagraphStem(ImageStem):
...
+        sharpness_kwargs=None,
+    ):
         super().__init__()
 
         if not augment:
             self.pil_transforms = transforms.Compose([transforms.CenterCrop(IMAGE_SHAPE)])
         else:
             if color_jitter_kwargs is None:
-                color_jitter_kwargs = {"brightness": (0.8, 1.6)}
+                color_jitter_kwargs = {"brightness": 0.4, "contrast": 0.4}
```
Comment (charlesfrye) [inline text_recognizer/stems/paragraph.py:36]: I feel like the jittering of the brightness and contrast could be turned up.

Reviewing examples from [the logged training predictions](https://wandb.ai/fullstackdeeplearning/fsdl-text-recognizer-2022-training/artifacts/run_table/run-2trhtbd4-trainpredictions/v98/files/train/predictions.table.json), it seems like the luminance levels are pretty uniform, and I suspect this is a big reason for our failure to generalize outside of the test distribution.

Comment (charlesfrye) [reply text_recognizer/stems/paragraph.py:36]: I would suggest playing around with these values in a tight loop in a notebook until the outputs look nicely diverse. DM me to schedule a pairing session if you want!

File: text_recognizer/stems/paragraph.py [hunk 4]
```diff
@@ -18,28 +18,51 @@ class ParagraphStem(ImageStem):
...
             if random_affine_kwargs is None:
                 random_affine_kwargs = {
-                    "degrees": 1,
-                    "shear": (-10, 10),
+                    "degrees": 3,
```
Comment (charlesfrye) [inline text_recognizer/stems/paragraph.py:39]: I think we could turn these down ever so slightly -- we're clipping the first few characters on about 5-10% of examples.

see example below, which is entry 15 [here](https://wandb.ai/fullstackdeeplearning/fsdl-text-recognizer-2022-training/artifacts/run_table/run-2trhtbd4-trainpredictions/v98/files/train/predictions.table.json), see also entry 3

image

File: text_recognizer/stems/paragraph.py [hunk 5]
```diff
@@ -18,28 +18,51 @@ class ParagraphStem(ImageStem):
...
+                    "scale": (0.95, 1.05),
                     "interpolation": transforms.InterpolationMode.BILINEAR,
                 }
+            if random_perspective_kwargs is None:
+                random_perspective_kwargs = {
+                    "distortion_scale": 0.2,
+                    "p": 0.8,
+                    "interpolation": transforms.InterpolationMode.BILINEAR,
+                }
+            if gaussian_blur_kwargs is None:
```
Comment (charlesfrye) [inline text_recognizer/stems/paragraph.py:50]: Good call on blurring -- I think this is something we'll see in production.

**Review by charlesfrye — CHANGES_REQUESTED** [review]
Nice work trying out some new augmentations.

I agree with most changes, but I think we should turn down the default values for the geometric transforms to avoid clipping (see comment). While we're at it, it'd be good to also turn up the default values for the luminance transforms.

File: text_recognizer/stems/paragraph.py (hunk 3 above)
Comment (charlesfrye) [reply text_recognizer/stems/paragraph.py:36]: But note that images logged on W&B are generated from _float tensors_, which means that there's an auto-rescale on the to get uint8 from 0-255.

Comment (charlesfrye) [reply text_recognizer/stems/paragraph.py:36]: (FROM SYNC MEETING) Current jitter is reducing the brightness of text more than it is increasing the brightness of the bg.

File: text_recognizer/stems/paragraph.py (hunk 4 above)
Comment (srbhchandra) [reply text_recognizer/stems/paragraph.py:39]: I stopped scaling up the images. It looks like it fixed the issue

File: text_recognizer/stems/paragraph.py (hunk 3 above)
Comment (srbhchandra) [reply text_recognizer/stems/paragraph.py:36]: didn't do much here

**Comment by charlesfrye** [comment]
see #41, which incorporates this PR

**Pull request closed without merging**